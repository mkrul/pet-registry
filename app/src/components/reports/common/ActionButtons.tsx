import React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faSave } from "@fortawesome/free-solid-svg-icons";
import Spinner from "../../common/Spinner";

interface ActionButtonsProps {
  isSaving: boolean;
  onSave: (e: React.FormEvent) => void;
  onCancel: () => void;
}

const ActionButtons: React.FC<ActionButtonsProps> = ({ isSaving, onSave, onCancel }) => {
  return (
    <div className="flex justify-end gap-2">
      <button
        type="submit"
        disabled={isSaving}
        className="px-4 py-2 bg-green-500 text-white rounded-md hover:bg-green-600 transition-colors disabled:opacity-50 disabled:cursor-not-allowed"
      >
        {isSaving ? (
          <div className="flex items-center">
            <Spinner inline size={16} className="mr-2" color="text-white" />
            Saving...
          </div>
        ) : (
          <>
            <FontAwesomeIcon icon={faSave} className="mr-2" />
            Save
          </>
        )}
      </button>
      <button
        type="button"
        onClick={onCancel}
        disabled={isSaving}
        className="px-4 py-2 border border-blue-500 bg-white text-blue-600 hover:bg-blue-100 rounded-md transition-colors disabled:opacity-50 disabled:cursor-not-allowed"
      >
        Cancel
      </button>
    </div>
  );
};

export default ActionButtons;
