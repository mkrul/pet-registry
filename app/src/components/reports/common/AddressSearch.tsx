import React, { useState } from "react";
import { Autocomplete, TextField, CircularProgress } from "@mui/material";
import { debounce } from "lodash";
import { commonInputStyles } from "../../../styles/commonStyles";

interface AddressSearchProps {
  onSearch: (address: string) => Promise<void>;
  disabled?: boolean;
}

interface AddressSuggestion {
  display_name: string;
  lat: string;
  lon: string;
}

const AddressSearch: React.FC<AddressSearchProps> = ({ onSearch, disabled }) => {
  const [inputValue, setInputValue] = useState("");
  const [options, setOptions] = useState<AddressSuggestion[]>([]);
  const [loading, setLoading] = useState(false);

  const fetchAddressSuggestions = React.useMemo(
    () =>
      debounce(async (input: string) => {
        if (input.length < 3) return;

        setLoading(true);
        try {
          const response = await fetch(
            `https://nominatim.openstreetmap.org/search?` +
              `format=json&` +
              `q=${encodeURIComponent(input)}` +
              `&countrycodes=us` +
              `&limit=5` +
              `&addressdetails=1`
          );
          const data = await response.json();
          setOptions(data);
        } catch (error) {
          console.error("Error fetching address suggestions:", error);
          setOptions([]);
        } finally {
          setLoading(false);
        }
      }, 300),
    []
  );

  React.useEffect(() => {
    if (inputValue) {
      fetchAddressSuggestions(inputValue);
    } else {
      setOptions([]);
    }
  }, [inputValue, fetchAddressSuggestions]);

  const handleOptionSelect = async (_: any, value: AddressSuggestion | null) => {
    if (value) {
      await onSearch(value.display_name);
    }
  };

  return (
    <div className="space-y-2">
      <label className="text-lg font-medium text-gray-900 mb-2">Address:</label>
      <Autocomplete
        fullWidth
        disabled={disabled}
        options={options}
        getOptionLabel={(option: AddressSuggestion) => option.display_name}
        filterOptions={x => x}
        loading={loading}
        onInputChange={(_, value) => setInputValue(value)}
        onChange={handleOptionSelect}
        noOptionsText="Start typing to search for addresses"
        loadingText="Loading suggestions..."
        ListboxProps={{
          style: { maxHeight: 200 }
        }}
        renderInput={params => (
          <TextField
            {...params}
            aria-label="Enter the last known address that the animal was seen at"
            InputProps={{
              ...params.InputProps,
              endAdornment: (
                <>
                  {loading ? <CircularProgress color="inherit" size={20} /> : null}
                  {params.InputProps.endAdornment}
                </>
              )
            }}
            sx={{
              ...commonInputStyles,
              backgroundColor: "white",
              "& .MuiOutlinedInput-root": {
                backgroundColor: "white"
              }
            }}
          />
        )}
      />
    </div>
  );
};

export default AddressSearch;
