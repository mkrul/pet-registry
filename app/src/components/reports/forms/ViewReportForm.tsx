import React, { useState } from "react";
import formatDate from "../../../lib/formatDate";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faPencil } from "@fortawesome/free-solid-svg-icons";
import Map from "../../common/Map";
import { ViewReportFormProps } from "../../../types/Report";
import Spinner from "../../common/Spinner";
import LocationDisplay from "../../common/LocationDisplay";
import DateDisplay from "../common/DateDisplay";
import { createMapLocation } from "../../../utils/mapUtils";
import { MAP_ZOOM_LEVELS } from "../../../constants/map";

const ViewReportForm: React.FC<ViewReportFormProps> = ({
  report,
  onEditClick,
  onBackClick,
  imageSrc,
  handleImageLoad,
  handleImageError
}) => {
  const [isImageLoading, setIsImageLoading] = useState(true);

  const onImageLoad = () => {
    setIsImageLoading(false);
    handleImageLoad();
  };

  return (
    <div className="space-y-6">
      {/* Action Buttons */}
      <div className="flex justify-end gap-2">
        <button
          type="button"
          onClick={onEditClick}
          className="px-4 py-2 bg-blue-500 text-white rounded-md hover:bg-blue-600 transition-colors"
        >
          <FontAwesomeIcon icon={faPencil} className="mr-2" />
          Edit
        </button>
        <button
          type="button"
          onClick={onBackClick}
          className="px-4 py-2 bg-gray-500 text-white rounded-md hover:bg-gray-600 transition-colors"
        >
          Back
        </button>
      </div>

      {/* Title */}
      <h2 className="text-2xl font-semibold text-gray-900">{report.title}</h2>

      {/* Image */}
      <div className="space-y-2">
        <div className="mt-1 relative">
          {isImageLoading && (
            <div className="absolute inset-0 flex items-center justify-center bg-gray-50 mb-4 mt-4">
              <Spinner size={32} />
            </div>
          )}
          <img
            src={imageSrc}
            alt={report.title}
            className="w-full rounded-lg shadow-sm"
            onLoad={onImageLoad}
            onError={handleImageError}
          />
        </div>
      </div>

      {/* Name */}
      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900 mb-2">Name:</label>
        <p className="text-md text-gray-500 mb-4">{report.name || "Unknown"}</p>
      </div>

      {/* Description */}
      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900 mb-2">Description:</label>
        <p className="text-gray-500 whitespace-pre-wrap">{report.description}</p>
      </div>

      {/* Species */}
      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900 mb-2">Species:</label>
        <p className="text-md text-gray-500 mb-4">{report.species}</p>
      </div>

      {/* Breeds */}
      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900 mb-2">Breed(s):</label>
        <div className="space-y-1">
          <p className="text-md text-gray-500">{report.breed1}</p>
          {report.breed2 && <p className="text-md text-gray-500">{report.breed2}</p>}
        </div>
      </div>

      {/* Gender */}
      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900 mb-2">Gender:</label>
        <p className="text-md text-gray-500 mb-4">{report.gender || "Unknown"}</p>
      </div>

      {/* Microchip ID */}
      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900 mb-2">Microchip ID:</label>
        <p className="text-md text-gray-500 mb-4">{report.microchipId || "Unknown"}</p>
      </div>

      {/* Altered Status */}
      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900 mb-2">Spayed/Neutered:</label>
        <p className="text-md text-gray-500 mb-4">
          {report.isAltered === true ? "Yes" : report.isAltered === false ? "No" : "Unknown"}
        </p>
      </div>

      {/* Colors */}
      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900 mb-2">Colors:</label>
        <div className="space-y-1">
          <p className="text-md text-gray-500">{report.color1}</p>
          {report.color2 && <p className="text-md text-gray-500">{report.color2}</p>}
          {report.color3 && <p className="text-md text-gray-500">{report.color3}</p>}
        </div>
      </div>

      {/* Location */}
      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900 mb-4">Location:</label>
        <LocationDisplay
          area={report.area}
          state={report.state}
          country={report.country}
          intersection={report.intersection}
        />
        <div className="mt-1">
          <Map
            initialLocation={createMapLocation(report)}
            readOnly={true}
            onLocationSelect={() => {}}
            initialZoom={MAP_ZOOM_LEVELS.VIEW}
          />
        </div>
      </div>
      <DateDisplay createdAt={report.createdAt} updatedAt={report.updatedAt} />
    </div>
  );
};

export default ViewReportForm;
