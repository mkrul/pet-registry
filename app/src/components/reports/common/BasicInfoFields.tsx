import React from "react";
import { TextField } from "@mui/material";
import { BasicInfoFieldsProps } from "../../../types/Report";
import { commonInputStyles } from "../../../styles/commonStyles";
import Tip from "../../common/Tip";
import { FormFieldError } from "../../common/FormFieldError";

export const BasicInfoFields: React.FC<BasicInfoFieldsProps> = ({
  formData,
  onInputChange,
  readOnly,
  error,
  descriptionError
}) => {
  return (
    <div className="space-y-6">
      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900 mb-2">Title:</label>
        <TextField
          name="title"
          value={formData.title || ""}
          onChange={onInputChange}
          variant="outlined"
          fullWidth
          required
          disabled={readOnly}
          sx={commonInputStyles}
          error={!!error}
        />
        <FormFieldError error={error} />
      </div>

      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900 mb-2">Pet's Name:</label>{" "}
        <span className="text-sm text-gray-500"> (Leave blank if not known)</span>
        <TextField
          name="name"
          value={formData.name || ""}
          onChange={onInputChange}
          variant="outlined"
          fullWidth
          disabled={readOnly}
          sx={commonInputStyles}
        />
      </div>

      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900 mb-2">Description:</label>
        <Tip>Describe the animal's appearance, behavior, and other relevant details.</Tip>
        <TextField
          name="description"
          value={formData.description || ""}
          onChange={onInputChange}
          multiline
          rows={4}
          variant="outlined"
          fullWidth
          required
          disabled={readOnly}
          sx={commonInputStyles}
          error={!!descriptionError}
        />
        <FormFieldError error={descriptionError} />
      </div>
    </div>
  );
};
