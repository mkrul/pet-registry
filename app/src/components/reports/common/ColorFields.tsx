import React from "react";
import { ColorFieldsProps } from "../../../types/Report";
import { AddFieldButton } from "../../common/AddFieldButton";
import { AdditionalFieldSet } from "../../common/AdditionalFieldSet";
import { ColorSearch } from "../../common/ColorSearch";
import { commonInputStyles } from "../../../styles/commonStyles";
import { Alert } from "@mui/material";

export const ColorFields: React.FC<ColorFieldsProps> = ({
  formData,
  showColor2,
  showColor3,
  setShowColor2,
  setShowColor3,
  onColor2Add,
  onColor3Add,
  onColor2Remove,
  onColor3Remove,
  isLoading,
  handleColor1Change,
  handleColor2Change,
  handleColor3Change,
  error
}) => {
  return (
    <div className="space-y-6">
      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900">Colors:</label>
        <div className="flex-grow">
          <ColorSearch
            value={formData.color1}
            onChange={handleColor1Change}
            disabled={isLoading}
            size="medium"
            excludeColors={[formData.color2, formData.color3].filter(Boolean)}
            sx={commonInputStyles}
            required
            error={!!error}
          />
          {error && (
            <Alert severity="error" sx={{ mb: 2 }}>
              {error}
            </Alert>
          )}
        </div>
      </div>

      {!showColor2 && formData.color1 && (
        <AddFieldButton
          onClick={() => setShowColor2(true)}
          disabled={isLoading}
          label="ADD COLOR"
          testId="add-color-button"
        />
      )}

      {showColor2 && (
        <div className="mt-8">
          <div className="mt-6">
            <AdditionalFieldSet
              label="Second Color:"
              onRemove={onColor2Remove}
              disabled={isLoading}
              testId="remove-color-button"
            >
              <ColorSearch
                value={formData.color2 || ""}
                onChange={handleColor2Change}
                disabled={isLoading}
                size="medium"
                excludeColors={[formData.color1, formData.color3].filter(Boolean)}
                sx={commonInputStyles}
              />
            </AdditionalFieldSet>
          </div>
        </div>
      )}

      {showColor2 && !showColor3 && formData.color2 && (
        <AddFieldButton
          onClick={() => setShowColor3(true)}
          disabled={isLoading}
          label="ADD COLOR"
          testId="add-color-button"
        />
      )}

      {showColor3 && (
        <div className="mt-8">
          <div className="mt-6">
            <AdditionalFieldSet
              label="Third Color:"
              onRemove={onColor3Remove}
              disabled={isLoading}
              testId="remove-color-button"
            >
              <ColorSearch
                value={formData.color3 || ""}
                onChange={handleColor3Change}
                disabled={isLoading}
                size="medium"
                excludeColors={[formData.color1, formData.color2].filter(Boolean)}
                sx={commonInputStyles}
              />
            </AdditionalFieldSet>
          </div>
        </div>
      )}
    </div>
  );
};
