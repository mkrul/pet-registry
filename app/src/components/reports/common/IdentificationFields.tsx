import React from "react";
import {
  FormControl,
  TextField,
  Select,
  MenuItem,
  RadioGroup,
  FormControlLabel,
  Radio
} from "@mui/material";
import { getGenderOptions } from "../../../lib/reports/genderList";
import speciesListJson from "../../../../../config/species.json";
import BreedSearch from "../../common/BreedSearch";
import { AddFieldButton } from "../../common/AddFieldButton";
import { AdditionalFieldSet } from "../../common/AdditionalFieldSet";
import { IdentificationFieldsProps } from "../../../types/Report";
import { commonInputStyles } from "../../../styles/commonStyles";
import { FormFieldError } from "../../common/FormFieldError";

export const IdentificationFields: React.FC<IdentificationFieldsProps> = ({
  formData,
  showBreed2,
  onInputChange,
  setShowBreed2,
  onBreed2Remove,
  isLoading,
  error,
  breedError
}) => {
  const genderOptions = getGenderOptions();
  const speciesOptions = speciesListJson.options;

  const createChangeEvent = (name: string, value: string) =>
    ({
      target: { name, value }
    }) as React.ChangeEvent<HTMLInputElement>;

  const handleFieldChange = (name: string) => (value: string) => {
    onInputChange(createChangeEvent(name, value));
  };

  const handleSpeciesChange = handleFieldChange("species");
  const handleBreedChange = handleFieldChange("breed1");
  const handleBreed2Change = handleFieldChange("breed2");
  const handleGenderChange = handleFieldChange("gender");

  const handleBreed2Remove = () => {
    onInputChange(createChangeEvent("breed2", ""));
    onBreed2Remove?.();
    setShowBreed2(false);
  };

  return (
    <div className="space-y-6">
      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900 mb-2">Microchip ID:</label>
        <span className="text-sm text-gray-500"> (Leave blank if not known)</span>
        <TextField
          data-testid="microchip-id-input"
          name="microchipId"
          value={formData.microchipId || ""}
          onChange={onInputChange}
          variant="outlined"
          fullWidth
          sx={commonInputStyles}
          disabled={isLoading}
        />
      </div>

      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900 mb-2">Gender:</label>{" "}
        <span className="text-sm text-gray-500"> (Leave blank if not known)</span>
        <FormControl fullWidth>
          <Select
            data-testid="gender-select"
            labelId="gender-label"
            id="gender"
            value={formData.gender || ""}
            onChange={e => handleGenderChange(e.target.value)}
            sx={commonInputStyles}
            disabled={isLoading}
            MenuProps={{
              PaperProps: {
                style: {
                  maxHeight: 200
                }
              }
            }}
          >
            {genderOptions.map((gender, index) => (
              <MenuItem key={index} value={gender} data-testid="gender-option">
                {gender}
              </MenuItem>
            ))}
          </Select>
        </FormControl>
      </div>

      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900">
          Is the animal spayed or neutered?
        </label>
        <div>
          <RadioGroup
            name="altered"
            value={formData.altered === null ? undefined : formData.altered.toString()}
            onChange={e => {
              console.log("Radio value changed to:", e.target.value);
              onInputChange(e);
            }}
          >
            <FormControlLabel value="1" control={<Radio />} label="Yes" />
            <FormControlLabel value="0" control={<Radio />} label="No" />
            <FormControlLabel value="" control={<Radio />} label="I don't know" />
          </RadioGroup>
          <FormFieldError error={error} />
        </div>
      </div>
      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900">Species:</label>
        <div>
          <TextField
            select
            name="species"
            value={formData.species || ""}
            onChange={onInputChange}
            variant="outlined"
            fullWidth
            required
            disabled={isLoading}
            sx={commonInputStyles}
            error={!!error}
          >
            {speciesOptions.map(option => (
              <MenuItem key={option} value={option}>
                {option}
              </MenuItem>
            ))}
          </TextField>
          <FormFieldError error={error} />
        </div>
      </div>

      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900">Breed:</label>
        <div>
          <BreedSearch
            species={formData.species.toLowerCase() as "dog" | "cat"}
            value={formData.breed1}
            onChange={handleBreedChange}
            disabled={isLoading}
            excludeBreeds={formData.breed2 ? [formData.breed2] : []}
            required
            size="medium"
            hideLabel
            disableClearable
            error={!!breedError}
            data-testid="breed-search"
          />
          <FormFieldError error={breedError} />
        </div>

        {!showBreed2 && formData.breed1 && (
          <AddFieldButton
            onClick={() => setShowBreed2(true)}
            disabled={isLoading}
            label="ADD BREED"
            testId="add-breed-button"
          />
        )}
      </div>

      {showBreed2 && (
        <div className="mt-8">
          <div className="mb-6">
            <AdditionalFieldSet
              label="Second Breed:"
              onRemove={handleBreed2Remove}
              disabled={isLoading}
              testId="remove-breed-button"
            >
              <BreedSearch
                species={formData.species.toLowerCase() as "dog" | "cat"}
                value={formData.breed2 || ""}
                onChange={handleBreed2Change}
                disabled={isLoading}
                excludeBreeds={[formData.breed1]}
                size="medium"
                hideLabel
                disableClearable
              />
            </AdditionalFieldSet>
          </div>
        </div>
      )}
    </div>
  );
};
