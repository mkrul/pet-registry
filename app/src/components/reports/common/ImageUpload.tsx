import React from "react";
import { Button, FormControl, Alert } from "@mui/material";
import { CloudUpload } from "@mui/icons-material";
import { ImageUploadProps } from "../../../types/common/Image";
import { commonInputStyles } from "../../../styles/commonStyles";
import Tip from "../../common/Tip";

export const ImageUpload: React.FC<ImageUploadProps> = ({
  onImageSelect,
  onFileChange,
  preview,
  disabled,
  onImageLoad,
  onImageError,
  error
}) => {
  const handleImageChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (e.target.files && e.target.files[0]) {
      const file = e.target.files[0];

      if (file.size > 5 * 1024 * 1024) {
        alert("File size exceeds 5MB.");
        return;
      }

      const validTypes = ["image/jpeg", "image/png", "image/gif"];
      if (!validTypes.includes(file.type)) {
        alert("Unsupported file type. Please upload a JPEG, PNG, or GIF image.");
        return;
      }

      const reader = new FileReader();
      reader.onloadend = () => {
        onImageSelect?.(file, reader.result as string);
        onFileChange?.(e);
      };
      reader.readAsDataURL(file);
    }
  };

  return (
    <div className="space-y-2">
      <label className="text-lg font-medium text-gray-900 mb-2">Photo:</label>
      <Tip>
        The best photos are taken outdoors with good lighting. Your photo should be well lit, with
        the animal's face centered in the frame.
      </Tip>
      <div className="mt-1">
        <div className="relative">
          <input
            type="file"
            name="image"
            accept="image/*"
            onChange={handleImageChange}
            style={{
              opacity: 0,
              position: "absolute",
              top: 0,
              left: 0,
              width: "100%",
              height: "100%",
              cursor: "pointer",
              zIndex: 1
            }}
            disabled={disabled}
            data-testid="file-input"
            required
          />
          <Button
            variant="outlined"
            startIcon={<CloudUpload />}
            disabled={disabled}
            sx={{
              ...commonInputStyles,
              width: "fit-content",
              position: "relative",
              borderColor: error ? "error.main" : undefined
            }}
          >
            Choose File
          </Button>
        </div>
        {error && (
          <Alert severity="error" sx={{ mb: 2 }}>
            {error}
          </Alert>
        )}
        {preview && (
          <div className="mt-2 relative w-48 h-48">
            <img
              src={preview}
              alt="Preview"
              className="object-cover w-full h-full rounded-md"
              onLoad={onImageLoad}
              onError={onImageError}
            />
          </div>
        )}
      </div>
    </div>
  );
};
