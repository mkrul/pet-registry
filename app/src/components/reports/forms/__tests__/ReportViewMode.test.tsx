import { describe, it, expect, vi } from "vitest";
import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/vitest";
import ViewReportForm from "../ViewReportForm";
import { ViewReportFormProps } from "../../../../types/Report";
import { ImageProps } from "../../../../types/common/Image";

vi.mock("../../../common/Map", () => ({
  default: () => <div data-testid="map">Map Component</div>
}));

vi.mock("../../../common/Spinner", () => ({
  default: () => <div data-testid="spinner">Loading...</div>
}));

vi.mock("../../../lib/formatDate", () => ({
  default: () => "12/31/2022, 7:00 PM"
}));

const mockReport = {
  id: 1,
  title: "Test Report",
  description: "Test Description",
  image: { variantUrl: "/test-image.jpg" } as ImageProps,
  species: "Dog",
  breed1: "Labrador",
  breed2: "Golden Retriever",
  color1: "Black",
  color2: "Brown",
  color3: "White",
  gender: "Male",
  name: "Max",
  microchipId: "123456",
  latitude: 0,
  longitude: 0,
  area: "Test Area",
  state: "Test State",
  country: "Test Country",
  createdAt: "2023-01-01",
  updatedAt: "2023-01-01",
  status: "active",
  archivedAt: null,
  recentlyUpdated: false,
  recentlyCreated: false
};

const mockProps: ViewReportFormProps = {
  report: mockReport,
  onEditClick: vi.fn(),
  onBackClick: vi.fn(),
  imageSrc: "/test-image.jpg",
  handleImageLoad: vi.fn(),
  handleImageError: vi.fn()
};

describe("ViewReportForm", () => {
  describe("Content Display", () => {
    it("renders basic report information", () => {
      const { container } = render(<ViewReportForm {...mockProps} />);

      expect(screen.getByText("Test Report")).toBeDefined();
      expect(screen.getByText("Test Description")).toBeDefined();
      expect(screen.getByText("Max")).toBeDefined();
      expect(screen.getByText("123456")).toBeDefined();
      expect(container).toMatchSnapshot();
    });

    it("renders pet details correctly", () => {
      render(<ViewReportForm {...mockProps} />);

      expect(screen.getByText("Dog")).toBeDefined();
      expect(screen.getByText("Labrador")).toBeDefined();
      expect(screen.getByText("Golden Retriever")).toBeDefined();
      expect(screen.getByText("Male")).toBeDefined();
    });

    it("renders all colors", () => {
      render(<ViewReportForm {...mockProps} />);

      expect(screen.getByText("Black")).toBeDefined();
      expect(screen.getByText("Brown")).toBeDefined();
      expect(screen.getByText("White")).toBeDefined();
    });

    it("renders location information", () => {
      render(<ViewReportForm {...mockProps} />);

      expect(screen.getByText("Test Area, Test State, Test Country")).toBeDefined();
      expect(screen.getByTestId("map")).toBeDefined();
    });

    it("renders dates correctly", () => {
      render(<ViewReportForm {...mockProps} />);

      const dates = screen.getAllByText("12/31/2022, 7:00 PM");
      expect(dates).toHaveLength(2);

      const postedAtLabel = screen.getByText("Posted at:");
      const updatedAtLabel = screen.getByText("Updated at:");

      expect(postedAtLabel).toBeDefined();
      expect(updatedAtLabel).toBeDefined();
      expect(postedAtLabel.nextElementSibling?.textContent).toBe("12/31/2022, 7:00 PM");
      expect(updatedAtLabel.nextElementSibling?.textContent).toBe("12/31/2022, 7:00 PM");
    });
  });

  describe("Image Handling", () => {
    it("renders image with loading state", () => {
      render(<ViewReportForm {...mockProps} />);

      const image = screen.getByRole("img");
      expect(image).toBeDefined();
      expect(image).toHaveAttribute("src", "/test-image.jpg");
      expect(screen.getByTestId("spinner")).toBeDefined();
    });

    it("handles image load event", () => {
      render(<ViewReportForm {...mockProps} />);

      const image = screen.getByRole("img");
      fireEvent.load(image);

      expect(mockProps.handleImageLoad).toHaveBeenCalled();
      expect(screen.queryByTestId("spinner")).toBeNull();
    });

    it("handles image error event", () => {
      render(<ViewReportForm {...mockProps} />);

      const image = screen.getByRole("img");
      fireEvent.error(image);

      expect(mockProps.handleImageError).toHaveBeenCalled();
    });
  });

  describe("Button Actions", () => {
    it("handles edit button click", () => {
      render(<ViewReportForm {...mockProps} />);

      const editButton = screen.getByText("Edit");
      fireEvent.click(editButton);

      expect(mockProps.onEditClick).toHaveBeenCalled();
    });

    it("handles back button click", () => {
      render(<ViewReportForm {...mockProps} />);

      const backButton = screen.getByText("Back");
      fireEvent.click(backButton);

      expect(mockProps.onBackClick).toHaveBeenCalled();
    });
  });

  describe("Edge Cases", () => {
    it("handles missing optional fields", () => {
      const partialReport = {
        ...mockReport,
        breed2: null,
        color2: null,
        color3: null,
        microchipId: null
      };

      render(<ViewReportForm {...mockProps} report={partialReport} />);

      expect(screen.getByText("Unknown")).toBeDefined();
      expect(screen.queryByText("Golden Retriever")).toBeNull();
    });
  });
});
