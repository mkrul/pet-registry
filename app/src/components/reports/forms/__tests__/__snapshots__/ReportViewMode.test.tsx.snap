// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ViewReportForm > Content Display > renders basic report information 1`] = `
<div>
  <div
    class="space-y-6"
  >
    <div
      class="flex justify-end mb-4 gap-2"
    >
      <button
        class="px-4 py-2 bg-blue-500 text-white rounded-md hover:bg-blue-600 transition-colors"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-pencil mr-2"
          data-icon="pencil"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M410.3 231l11.3-11.3-33.9-33.9-62.1-62.1L291.7 89.8l-11.3 11.3-22.6 22.6L58.6 322.9c-10.4 10.4-18 23.3-22.2 37.4L1 480.7c-2.5 8.4-.2 17.5 6.1 23.7s15.3 8.5 23.7 6.1l120.3-35.4c14.1-4.2 27-11.8 37.4-22.2L387.7 253.7 410.3 231zM160 399.4l-9.1 22.7c-4 3.1-8.5 5.4-13.3 6.9L59.4 452l23-78.1c1.4-4.9 3.8-9.4 6.9-13.3l22.7-9.1 0 32c0 8.8 7.2 16 16 16l32 0zM362.7 18.7L348.3 33.2 325.7 55.8 314.3 67.1l33.9 33.9 62.1 62.1 33.9 33.9 11.3-11.3 22.6-22.6 14.5-14.5c25-25 25-65.5 0-90.5L453.3 18.7c-25-25-65.5-25-90.5 0zm-47.4 168l-144 144c-6.2 6.2-16.4 6.2-22.6 0s-6.2-16.4 0-22.6l144-144c6.2-6.2 16.4-6.2 22.6 0s6.2 16.4 0 22.6z"
            fill="currentColor"
          />
        </svg>
        Edit
      </button>
      <button
        class="px-4 py-2 bg-gray-500 text-white rounded-md hover:bg-gray-600 transition-colors"
        type="button"
      >
        Back
      </button>
    </div>
    <h2
      class="text-2xl font-semibold text-gray-900"
    >
      Test Report
    </h2>
    <div
      class="space-y-2"
    >
      <div
        class="mt-1 relative"
      >
        <div
          class="absolute inset-0 flex items-center justify-center bg-gray-50 mb-4 mt-4"
        >
          <div
            data-testid="spinner"
          >
            Loading...
          </div>
        </div>
        <img
          alt="Test Report"
          class="w-full rounded-lg shadow-sm"
          src="/test-image.jpg"
        />
      </div>
    </div>
    <div
      class="space-y-2"
    >
      <label
        class="text-lg font-medium text-gray-900 mb-2"
      >
        Description:
      </label>
      <p
        class="text-gray-500 whitespace-pre-wrap"
      >
        Test Description
      </p>
    </div>
    <div
      class="space-y-2"
    >
      <label
        class="text-lg font-medium text-gray-900 mb-2"
      >
        Name:
      </label>
      <p
        class="text-md text-gray-500 mb-4"
      >
        Max
      </p>
    </div>
    <div
      class="space-y-2"
    >
      <label
        class="text-lg font-medium text-gray-900 mb-2"
      >
        Species:
      </label>
      <p
        class="text-md text-gray-500 mb-4"
      >
        Dog
      </p>
    </div>
    <div
      class="space-y-2"
    >
      <label
        class="text-lg font-medium text-gray-900 mb-2"
      >
        Breed(s):
      </label>
      <div
        class="space-y-1"
      >
        <p
          class="text-md text-gray-500"
        >
          Labrador
        </p>
        <p
          class="text-md text-gray-500"
        >
          Golden Retriever
        </p>
      </div>
    </div>
    <div
      class="space-y-2"
    >
      <label
        class="text-lg font-medium text-gray-900 mb-2"
      >
        Gender:
      </label>
      <p
        class="text-md text-gray-500 mb-4"
      >
        Male
      </p>
    </div>
    <div
      class="space-y-2"
    >
      <label
        class="text-lg font-medium text-gray-900 mb-2"
      >
        Microchip ID:
      </label>
      <p
        class="text-md text-gray-500 mb-4"
      >
        123456
      </p>
    </div>
    <div
      class="space-y-2"
    >
      <label
        class="text-lg font-medium text-gray-900 mb-2"
      >
        Colors:
      </label>
      <div
        class="space-y-1"
      >
        <p
          class="text-md text-gray-500"
        >
          Black
        </p>
        <p
          class="text-md text-gray-500"
        >
          Brown
        </p>
        <p
          class="text-md text-gray-500"
        >
          White
        </p>
      </div>
    </div>
    <div
      class="space-y-2"
    >
      <label
        class="text-lg font-medium text-gray-900 mb-4"
      >
        Location:
      </label>
      <p
        class="text-gray-500 mt-2"
      >
        Test Area, Test State, Test Country
      </p>
      <div
        class="mt-1"
      >
        <div
          data-testid="map"
        >
          Map Component
        </div>
      </div>
    </div>
    <div
      class="space-y-2"
    >
      <div
        class="flex gap-8"
      >
        <div>
          <label
            class="text-lg font-medium text-gray-900 mb-2"
          >
            Posted at:
          </label>
          <p
            class="text-md text-gray-500 mb-4"
          >
            12/31/2022, 7:00 PM
          </p>
        </div>
        <div>
          <label
            class="text-lg font-medium text-gray-900 mb-2"
          >
            Updated at:
          </label>
          <p
            class="text-md text-gray-500 mb-4"
          >
            12/31/2022, 7:00 PM
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
